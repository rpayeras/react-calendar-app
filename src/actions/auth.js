import Swal from "sweetalert2";
import { fetchNoToken, fetchWithToken } from "../helpers/fetch";
import { types } from "../types/types";

export const startLogin = (email, password) => {
    return async (dispatch) => {
        try{
            const res = await fetchNoToken('auth', {email, password}, 'POST')
            const {ok, data, msg} = await res.json()

            if(ok){
                const {name, uid, token} = data;

                localStorage.setItem('token', token)
                localStorage.setItem('token-created', new Date().getTime())
                
                dispatch(login({
                    uid,
                    name
                }))
            } else {
                Swal.fire('Error',msg, 'error')
            }
        } catch(err){
            Swal.fire('Error', err, 'error')
        }
    }
}

export const startRegister = (name, email, password) => {
    return async (dispatch) => {
        try{
            const res = await fetchNoToken('auth/new', {name, email, password}, 'POST')
            const {ok, data, msg, errors} = await res.json()
            let errMsg = '';
            
            if(errors){
                errors.forEach(({msg}, idx) => {
                    errMsg += `\n${idx}: ${msg} \n`
                })
            }

            if(ok){
                const {name, uid, token} = data;

                localStorage.setItem('token', token)
                localStorage.setItem('token-created', new Date().getTime())
                
                dispatch(register({
                    uid,
                    name
                }))
            } else {
                Swal.fire('Error', errMsg, 'error')
            }
        } catch(err){
            Swal.fire('Error', err, 'error')
        }
    }
}

export const startChecking = () => {
    return async (dispatch) => {
        try{
            const res = await fetchWithToken('auth/renew')
            const {ok, data, msg} = await res.json()

            if(ok){
                const {name, uid, token} = data;

                localStorage.setItem('token', token)
                localStorage.setItem('token-created', new Date().getTime())
                
                dispatch(login({
                    uid,
                    name
                }))
            } else {
                console.log(msg)
                //Swal.fire('Error', msg, 'error')
            }

            dispatch(checkingFinish())
        } catch(err){
            console.log(err)
            dispatch(checkingFinish())
            // Swal.fire('Error', err, 'error')
        }
    }
}

export const startLogout = () => {
    return (dispatch) => {
        localStorage.clear()
        dispatch(logout())
    }
}

const checkingFinish = () => ({ type: types.authCheckingFinish })

export const login = (user) => ({
    type: types.authLogin,
    payload: user
})

export const register = (user) => ({
    type: types.authRegister,
    payload: user
})

export const logout = () => ({
    type: types.authLogout
})
