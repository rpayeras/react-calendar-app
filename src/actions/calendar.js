import { types } from "../types/types"
import {fetchWithToken} from "../helpers/fetch"
import { prepareEvents } from "../helpers/prepareEvents";
import Swal from "sweetalert2";

export const eventStartLoading = () => {
    return async (dispatch) => {

        try{
            const res = await fetchWithToken('events')
            const {ok, data} = await res.json()


            if(ok){
                let events = prepareEvents(data)
                dispatch(eventLoaded(events))
            }
        } catch(err) {
            console.log(err)
        }
    }
}

const eventLoaded = (events) => ({
    type: types.eventLoaded,
    payload: events
})

export const eventStartAddNew = (event) => {
    return async(dispatch, getState) => {
        const {uid, name} = getState().auth;

        try{
            const res = await fetchWithToken('events', event, 'POST')
            const {ok, data} = await res.json();

            if(ok){
                event.id = data.id;
                event.user = {
                    _id: uid,
                    name
                }

                dispatch(eventAddNew(event))
            }
        } catch (err) {
            console.log(err);
        }
    }
}

export const eventAddNew = (event) => ({
    type: types.eventAddNew,
    payload: event
})

export const eventSetCurrent = (event) => ({
    type: types.eventSetCurrent,
    payload: event
});

export const eventCleanCurrent = (event) => ({
    type: types.eventCleanCurrent,
});

export const eventStartUpdateCurrent = (event) => {
    return async(dispatch) => {
        try{
            const res = await fetchWithToken(`events/${event.id}`, event, 'PUT')
            const body = await res.json()

            if(body.ok){
                dispatch(eventUpdateCurrent(event))
            } else {
                Swal.fire('Error', body.msg, 'error')
            }
        } catch(err){
            console.log(err)
        }
    }
}

export const eventUpdateCurrent = (event) => ({
    type: types.eventUpdateCurrent,
    payload: event
});

export const eventStartDelete = () => {
    return async(dispatch, getState) => {
        const {id} = getState().calendar.current
        
        try{
            const res = await fetchWithToken(`events/${id}`, {}, 'DELETE')
            const body = await res.json()

            if(body.ok){
                dispatch(eventDelete())
            } else {
                Swal.fire('Error', body.msg, 'error')
            }
        } catch(err){
            console.log(err)
        }
    }
}

export const eventDelete = () => ({
    type: types.eventDelete
})

export const eventLogout = () => ({
    type: types.eventLogout
})