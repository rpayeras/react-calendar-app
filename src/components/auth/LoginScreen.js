import React from 'react';
import './login.css';

import {useForm} from '../../hooks/useForm'
import { useDispatch } from 'react-redux';
import { startLogin, startRegister } from '../../actions/auth';
import Swal from 'sweetalert2';

export const LoginScreen = () => {
    const [formLoginValues, setLoginFormValue] = useForm({
        lEmail: "email12@test.com",
        lPassword: "1234test"
    })

    const dispatch = useDispatch()

    const {lEmail, lPassword} = formLoginValues;

    const [formRegisterValues, setRegisterFormValue] = useForm({
        rName: "User",
        rEmail: "email999@test.com",
        rPassword: "1234test",
        rPassword2: "1234test"
    })

    const {rName, rEmail, rPassword, rPassword2} = formRegisterValues;

    const handleLogin = (e) => {
        e.preventDefault()
        
        dispatch(startLogin(lEmail, lPassword))
    }

    const handleRegister = (e) => {
        e.preventDefault()

        if(rPassword !== rPassword2){
            return Swal.fire('Error', 'Passwords must be the same', 'error')
        }

        dispatch(startRegister(rName, rEmail, rPassword))
        
        console.log(formRegisterValues);
    }

    
    return (
        <div className="container login-container">
            <div className="row">
                <div className="col-md-6 login-form-1">
                    <h3>Sign In</h3>
                    <form onSubmit={handleLogin}>
                        <div className="form-group">
                            <input 
                                type="text"
                                className="form-control"
                                placeholder="Email"
                                name="lEmail"
                                value={lEmail}
                                onChange={setLoginFormValue}
                            />
                        </div>
                        <div className="form-group">
                            <input
                                type="password"
                                className="form-control"
                                placeholder="Password"
                                name="lPassword"
                                value={lPassword}
                                onChange={setLoginFormValue}
                            />
                        </div>
                        <div className="form-group">
                            <input 
                                type="submit"
                                className="btnSubmit"
                                value="Login" 
                            />
                        </div>
                    </form>
                </div>

                <div className="col-md-6 login-form-2">
                    <h3>Sign up</h3>
                    <form onSubmit={handleRegister}>
                        <div className="form-group">
                            <input
                                type="text"
                                className="form-control"
                                placeholder="Name"
                                name="rName"
                                value={rName}
                                onChange={setRegisterFormValue}
                            />
                        </div>
                        <div className="form-group">
                            <input
                                type="email"
                                className="form-control"
                                placeholder="Email"
                                name="rEmail"
                                value={rEmail}
                                onChange={setRegisterFormValue}
                            />
                        </div>
                        <div className="form-group">
                            <input
                                type="password"
                                className="form-control"
                                placeholder="Password" 
                                name="rPassword"
                                value={rPassword}
                                onChange={setRegisterFormValue}
                            />
                        </div>

                        <div className="form-group">
                            <input
                                type="password"
                                className="form-control"
                                placeholder="Repeat password" 
                                name="rPassword2"
                                value={rPassword2}
                                onChange={setRegisterFormValue}
                            />
                        </div>

                        <div className="form-group">
                            <input 
                                type="submit" 
                                className="btnSubmit" 
                                value="Create account" />
                        </div>
                    </form>
                </div>
            </div>
        </div>
    )
}