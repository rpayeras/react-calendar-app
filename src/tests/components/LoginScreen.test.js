import React from "react";
import { mount } from "enzyme";
import { Provider } from "react-redux";

import configureStore from "redux-mock-store";
import thunk from "redux-thunk";

import { LoginScreen } from "../../components/auth/LoginScreen";
import { MemoryRouter } from "react-router";
import { createMemoryHistory } from "history";
import { startLogin, startRegister } from "../../actions/auth";
import Swal from "sweetalert2";

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

const initState = {
  auth: {
    checking: false,
  },
};

let store = mockStore(initState);
const history = createMemoryHistory();

store.dispatch = jest.fn();

const wrapper = mount(
  <Provider store={store}>
    <LoginScreen />
  </Provider>
);

jest.mock("../../actions/auth", () => ({
  startLogin: jest.fn(),
  startRegister: jest.fn(),
}));

Swal.fire = jest.fn();

describe("Testing LoginScreen component", () => {
  beforeEach(() => {
    store = mockStore(initState);
    jest.clearAllMocks();
  });

  test("should render", () => {
    expect(wrapper).toMatchSnapshot();
  });

  test("should call dispatch login", () => {
    wrapper.find('input[name="lEmail"]').simulate("change", {
      target: {
        name: "lEmail",
        value: "email12@test.com",
      },
    });

    wrapper.find('input[name="lPassword"]').simulate("change", {
      target: {
        name: "lPassword",
        value: "123456",
      },
    });

    wrapper.find("form").at(0).prop("onSubmit")({
      preventDefault() {},
    });

    expect(startLogin).toHaveBeenCalledWith("email12@test.com", "123456");
  });

  test("should call dispatch register", () => {
    wrapper.find('input[name="rName"]').simulate("change", {
      target: {
        name: "rName",
        value: "User",
      },
    });

    wrapper.find('input[name="rEmail"]').simulate("change", {
      target: {
        name: "rEmail",
        value: "email12@test.com",
      },
    });

    wrapper.find('input[name="rPassword"]').simulate("change", {
      target: {
        name: "rPassword",
        value: "123456",
      },
    });

    wrapper.find('input[name="rPassword2"]').simulate("change", {
      target: {
        name: "rPassword2",
        value: "123456",
      },
    });

    wrapper.find("form").at(1).prop("onSubmit")({
      preventDefault() {},
    });

    expect(startRegister).toHaveBeenCalledWith(
      "User",
      "email12@test.com",
      "123456"
    );
  });

  test("should detect passwords are not same and show alert", () => {
    wrapper.find('input[name="rName"]').simulate("change", {
      target: {
        name: "rName",
        value: "User",
      },
    });

    wrapper.find('input[name="rEmail"]').simulate("change", {
      target: {
        name: "rEmail",
        value: "email12@test.com",
      },
    });

    wrapper.find('input[name="rPassword"]').simulate("change", {
      target: {
        name: "rPassword",
        value: "123456",
      },
    });

    wrapper.find('input[name="rPassword2"]').simulate("change", {
      target: {
        name: "rPassword",
        value: "32424222",
      },
    });

    wrapper.find("form").at(1).prop("onSubmit")({
      preventDefault() {},
    });

    expect(Swal.fire).toHaveBeenCalledWith(
      "Error",
      "Passwords must be the same",
      "error"
    );
    expect(startRegister).not.toHaveBeenCalled();
  });
});
