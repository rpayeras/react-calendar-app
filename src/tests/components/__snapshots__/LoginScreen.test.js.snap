// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing LoginScreen component should render 1`] = `
<div
  className="container login-container"
>
  <div
    className="row"
  >
    <div
      className="col-md-6 login-form-1"
    >
      <h3>
        Sign In
      </h3>
      <form
        onSubmit={[Function]}
      >
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="lEmail"
            onChange={[Function]}
            placeholder="Email"
            type="text"
            value=""
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="lPassword"
            onChange={[Function]}
            placeholder="Password"
            type="password"
            value=""
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="btnSubmit"
            type="submit"
            value="Login"
          />
        </div>
      </form>
    </div>
    <div
      className="col-md-6 login-form-2"
    >
      <h3>
        Sign up
      </h3>
      <form
        onSubmit={[Function]}
      >
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="rName"
            onChange={[Function]}
            placeholder="Name"
            type="text"
            value=""
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="rEmail"
            onChange={[Function]}
            placeholder="Email"
            type="email"
            value=""
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="rPassword"
            onChange={[Function]}
            placeholder="Password"
            type="password"
            value=""
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="rPassword2"
            onChange={[Function]}
            placeholder="Repeat password"
            type="password"
            value=""
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="btnSubmit"
            type="submit"
            value="Create account"
          />
        </div>
      </form>
    </div>
  </div>
</div>
`;
